# QTM 385 - Experimental Methods

Assignment 05

# Instructions

As in previous assignments, here are 10 questions related to our recent
classes. They cover one-sided and two-sided non-compliance. Please
answer all question and export the file as a PDF or html file. If you
use R or Jupyter Notebook, make sure to include both the code and the
output in the document.

If you have any questions about the assignment, feel free to email me at
<danilo.freire@emory.edu>.

Good luck!

# Questions

1.  Name the four compliance types in two-sided non-compliance, and
    provide a brief description of each. Imagine a scenario in which
    each type of non-compliance could occur.

In a two-sided non-compliance scenario, individuals may or may not adhere to their assigned treatment, leading to four compliance types. Compliers follow their assignment—those in the treatment group receive the treatment, while those in the control group do not seek it elsewhere. For example, in a school tutoring program, a student assigned to receive tutoring attends all sessions, while a student in the control group does not seek extra help. Always-Takers receive the treatment regardless of their assignment, such as a student in a mental health study who seeks therapy independently even when placed in the control group. Never-Takers refuse the treatment regardless of assignment, like a patient in a clinical trial who declines medication even when assigned to the treatment group. Defiers do the opposite of their assignment—those in the treatment group refuse the treatment, while those in the control group actively seek it elsewhere. An example would be a workplace wellness program where employees assigned to the program refuse to participate, while those in the control group enroll in similar external programs on their own.

2.  Explain the concept of Two-Stage Least Squares (2SLS) in the context
    of estimating CACE. Briefly describe the two stages.

Two-Stage Least Squares (2SLS) is a statistical method used to estimate the Complier Average Causal Effect (CACE) in the presence of non-compliance. It accounts for endogeneity by using an instrumental variable (IV) approach, where treatment assignment serves as an instrument for actual treatment received. In the first stage, the treatment received is regressed on the randomly assigned treatment to predict the probability of receiving treatment based solely on assignment. This helps isolate the variation in treatment that is exogenous. In the second stage, the outcome variable is regressed on the predicted treatment values from the first stage, allowing for an unbiased estimate of the causal effect of treatment on compliers. This method effectively removes bias from non-compliance and ensures that the estimated effect reflects the impact on those who adhere to their assignment.

3.  Discuss the trade-off between internal validity and external
    validity when focusing on CACE rather than ATE in the presence of
    non-compliance.

Focusing on the Complier Average Causal Effect (CACE) rather than the Average Treatment Effect (ATE) in the presence of non-compliance involves a trade-off between internal validity and external validity. CACE provides a more internally valid estimate of the treatment effect for compliers by accounting for non-compliance and ensuring that the estimated effect is not biased by individuals who defy their assignment. This makes it a more precise measure of causality within the study population. However, CACE estimates only apply to compliers, a specific subgroup, which limits external validity since the effect may not generalize to always-takers, never-takers, or defiers in other settings. In contrast, ATE seeks to estimate the average effect for the entire population, including non-compliers, making it more generalizable but potentially less internally valid due to bias from selection effects. The trade-off is that while CACE provides a cleaner causal estimate, it may not apply broadly to all individuals, whereas ATE gives a more comprehensive picture at the cost of potential bias from non-compliance.

4.  The following three quantities are similar in appearance but refer
    to different things. Describe the differences.

$$
E[Y_i(d_i(1)) | D_i = 1]
$$

$$
E[Y_i(d_i(1)) | d_i(1) = 1]
$$

$$
E[Y_i(d_i(1)) | d_i(1) = d_i(0) = 1]
$$

The three quantities differ based on the conditioning variables and the subgroup they describe. The first expression, \( E[Y_i(d_i(1)) | D_i = 1] \), represents the expected outcome under treatment for individuals assigned to treatment, regardless of whether they comply. This mixes compliers, always-takers, and defiers, making it potentially biased by non-compliance. The second expression, \( E[Y_i(d_i(1)) | d_i(1) = 1] \), conditions on actually receiving treatment, regardless of assignment. This quantity includes both compliers and always-takers but excludes never-takers and defiers, meaning it does not estimate a causal effect strictly based on random assignment. The third expression, \( E[Y_i(d_i(1)) | d_i(1) = d_i(0) = 1] \), conditions on individuals who always take the treatment whether assigned to it or not (always-takers). Since their treatment status does not depend on random assignment, this expression does not reflect a causal effect of the treatment but rather an association specific to this subgroup. The key distinction among the three expressions is how they define the subset of individuals being considered, influencing whether they provide an internally valid estimate of treatment effects.

1.  Imagine we have a dataset for a hypothetical experiment evaluating a
    “Financial Literacy Programme” aimed at improving personal financial
    management skills. The dataset is available at
    <https://github.com/danilofreire/qtm385/blob/main/assignments/05-assignment/financial.csv>.
    The dataset has the following variables:

-   `id`: Unique participant identifier.
-   `assignment`: Treatment assignment (0 = control group, 1 = treatment
    group). Participants assigned to the treatment group were offered
    access to the programme.
-   `programme_Usage`: Actual usage of the Financial Literacy Programme
    (0 = no usage, 1 = used the programme). This indicates whether
    participants in either group actually used the programme
    (treatment). Note that control group members could theoretically
    access similar resources outside the study, leading to potential
    two-sided non-compliance aspects, though we will primarily focus on
    IV analysis suitable for non-compliance scenarios discussed in
    lectures.
-   `financial_knowledge_score`: Score on a financial knowledge test
    post-experiment (outcome variable). Higher scores indicate better
    financial knowledge.

Estimate the intent-to-treat (ITT) effect of `assignment` on
`financial_knowledge_score` using robust linear regression. Interpret,
in detail, the ITT effect in the context of the Financial Literacy
Programme.

In [1]:
import pandas as pd
import statsmodels.api as sm

# Load the dataset using a relative file path
df = pd.read_csv("financial.csv")  # Since the file is in the same folder as the notebook

# Define independent and dependent variables
X = df['assignment']  # Treatment assignment (0 = control, 1 = treatment)
y = df['financial_knowledge_score']  # Outcome variable

# Add a constant term for the regression model
X = sm.add_constant(X)

# Run robust linear regression
model = sm.OLS(y, X).fit(cov_type='HC3')  # Using HC3 robust standard errors

# Print regression results
print(model.summary())


                                OLS Regression Results                               
Dep. Variable:     financial_knowledge_score   R-squared:                       0.620
Model:                                   OLS   Adj. R-squared:                  0.616
Method:                        Least Squares   F-statistic:                     156.8
Date:                       Mon, 24 Feb 2025   Prob (F-statistic):           4.75e-22
Time:                               15:38:17   Log-Likelihood:                -439.60
No. Observations:                        100   AIC:                             883.2
Df Residuals:                             98   BIC:                             888.4
Df Model:                                  1                                         
Covariance Type:                         HC3                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------

The estimated ITT effect of assignment on financial knowledge score is 50.16, with a standard error of 4.006, and is highly statistically significant (p < 0.001). This means that being assigned to the Financial Literacy Programme leads to a substantial increase in financial knowledge scores on average.

This effect measures the impact of being offered the programme, regardless of actual participation. Since some assigned individuals may not use the programme, and some in the control group might find similar resources, ITT accounts for real-world compliance issues. The 50-point increase suggests that access to financial literacy resources significantly improves knowledge, possibly due to increased awareness or motivation. The R-squared of 0.62 indicates assignment explains much of the score variation. ITT provides a policy-relevant estimate, showing the likely effect if such a programme were widely implemented, though it does not isolate the effect on actual users, which would require a CACE analysis.


1.  Estimate the ITT effect of `assignment` on `programme_usage` using
    robust linear regression. Interpret this ITT_D in terms of programme
    uptake.

In [2]:
# Define independent and dependent variables
X = df['assignment']  # Treatment assignment (0 = control, 1 = treatment)
y = df['programme_usage']  # Actual programme usage (0 = no usage, 1 = used programme)

# Add a constant term for the regression model
X = sm.add_constant(X)

# Run robust linear regression
model = sm.OLS(y, X).fit(cov_type='HC3')  # Using HC3 robust standard errors

# Print regression results
print(model.summary())


                            OLS Regression Results                            
Dep. Variable:        programme_usage   R-squared:                       0.532
Model:                            OLS   Adj. R-squared:                  0.527
Method:                 Least Squares   F-statistic:                     109.2
Date:                Mon, 24 Feb 2025   Prob (F-statistic):           1.29e-17
Time:                        15:43:16   Log-Likelihood:                -33.316
No. Observations:                 100   AIC:                             70.63
Df Residuals:                      98   BIC:                             75.84
Df Model:                           1                                         
Covariance Type:                  HC3                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0600      0.034      1.751      0.0

The ITT estimate of 0.72 for assignment on programme usage indicates that being assigned to the Financial Literacy Programme increases the likelihood of actual participation by 72 percentage points on average. This effect is highly significant (p < 0.001), meaning the programme assignment strongly influences uptake. However, the coefficient is less than 1, suggesting some level of non-compliance—some participants assigned to the programme did not use it, and some in the control group may have accessed similar resources. The relatively high R-squared value of 0.532 implies that treatment assignment explains a substantial portion of the variation in programme usage.


2.  Use the IV approach to estimate the CACE. Interpret the estimates
    and compare with the results from the ITT analysis you conducted in
    questions 5 and 6.

In [4]:
from linearmodels.iv import IV2SLS  # For Instrumental Variables (IV) regression

# Define variables
Z = df['assignment']  # Instrumental variable (random assignment)
X = df['programme_usage']  # Treatment actually received
Y = df['financial_knowledge_score']  # Outcome variable

# Add a constant term
df['const'] = 1

# Run IV regression: First stage (programme_usage ~ assignment), Second stage (financial_knowledge_score ~ programme_usage)
iv_model = IV2SLS(dependent=Y, exog=df[['const']], endog=X, instruments=Z).fit(cov_type="robust")

# Print the results
print(iv_model.summary)


                              IV-2SLS Estimation Summary                             
Dep. Variable:     financial_knowledge_score   R-squared:                      0.5786
Estimator:                           IV-2SLS   Adj. R-squared:                 0.5743
No. Observations:                        100   F-statistic:                    147.14
Date:                       Mon, Feb 24 2025   P-value (F-stat)                0.0000
Time:                               16:28:58   Distribution:                  chi2(1)
Cov. Estimator:                       robust                                         
                                                                                     
                                Parameter Estimates                                
                 Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
-----------------------------------------------------------------------------------
const               35.580     3.0694     11.592     0.0000 

The IV-2SLS estimation results provide the Complier Average Causal Effect (CACE) of programme usage on financial knowledge scores. The estimated effect is 69.67 with a standard error of 5.74, and it is highly significant (p < 0.001). This means that for individuals who actually used the Financial Literacy Programme due to their assignment (compliers), their financial knowledge score increased by approximately 69.67 points on average. This estimate is larger than the ITT effect of assignment on financial knowledge (50.16), which includes both compliers and non-compliers, diluting the effect of actual programme participation.

Comparing with previous ITT estimates, the ITT effect of assignment on programme usage was 0.72, meaning assignment increased the likelihood of participation by 72 percentage points. The ratio of ITT_Y to ITT_D aligns with the CACE estimate, as 50.16 divided by 0.72 is approximately 69.67. This confirms that the IV approach successfully isolates the effect of programme usage among compliers. The R-squared value of 0.5786 indicates that programme usage explains a significant portion of the variation in financial knowledge scores. Overall, the IV results show that for those who actually used the programme due to their assignment, the financial knowledge benefits were even greater than the ITT estimate suggested.

3.  In her study of election monitoring in Indonesia, [Hyde
    (2010)](https://doi.org/10.1017/S1537592710001222) randomly assigned
    international election observers to monitor certain polling
    stations. Here, we consider a subset of her experiment where
    approximately 20% of the villages were assigned to the treatment
    group. Because of difficult terrain and time constraints, observers
    monitored 68 of the 409 polling places assigned to treatment.
    Observers also monitored 21 of the 1,562 stations assigned to the
    control group. The dependent variable here is the number of ballots
    that were declared invalid by polling station officials.

Explain what the non-interference assumption means in the context of
this experiment.

The non-interference assumption in this experiment means that the assignment of election observers to certain polling stations should not affect the outcomes at other polling stations. In other words, the presence of observers at a treated polling station should not influence the number of invalid ballots at control stations. This assumption is crucial for ensuring that the estimated effect of election monitoring is not biased by spillover effects, such as election officials changing their behavior at unmonitored stations due to knowledge that other stations are being observed. If interference occurs, it would violate the assumption that the treatment affects only the treated units, making it difficult to isolate the true causal impact of election monitoring.

In the context of this study, potential violations of the non-interference assumption could arise if election officials or political actors respond strategically to the presence of observers. For example, if officials at unmonitored stations adjust their behavior to align with what they expect is happening at monitored stations, the control group’s outcomes may no longer reflect a true counterfactual. Additionally, if information spreads among voters or local officials, influencing decisions on ballot validation or fraud at control stations, this could also lead to interference. Ensuring minimal interaction between treatment and control stations is essential for maintaining the validity of causal estimates in this experiment.

1.  Download the sample dataset at
    <https://github.com/danilofreire/qtm385/blob/main/assignments/05-assignment/hyde.csv>
    and estimate the ITT and the CACE. Interpret the results.

In [6]:

# Load dataset (replace 'your_dataset.csv' with the actual file path)
df = pd.read_csv("hyde.csv")

# Compute ITT: Difference in means of invalid_ballots between treatment (sample=1) and control (sample=0)
itt = df[df["sample"] == 1]["invalid_ballots"].mean() - df[df["sample"] == 0]["invalid_ballots"].mean()

# Compute proportion of observed taking treatment in both groups
p_treated_given_sample_1 = df[df["sample"] == 1]["observed"].mean()  # P(D=1 | Z=1)
p_treated_given_sample_0 = df[df["sample"] == 0]["observed"].mean()  # P(D=1 | Z=0)

# Compute CACE
cace = itt / (p_treated_given_sample_1 - p_treated_given_sample_0)

# Print results
print(f"Intent-to-Treat (ITT): {itt:.4f}")
print(f"Complier Average Causal Effect (CACE): {cace:.4f}")


Intent-to-Treat (ITT): 4.8241
Complier Average Causal Effect (CACE): 31.5682


2.  Use randomisation inference to test the sharp null hypothesis that
    there is no intent-to-treat effect for any polling location.
    Interpret the results. Explain why testing the null hypothesis that
    the ITT is zero for all subjects serves the same purpose as testing
    the null hypothesis that the ATE is zero for all Compliers.

In [9]:
import numpy as np
import pandas as pd

# Compute the observed ITT
observed_itt = df[df["sample"] == 1]["invalid_ballots"].mean() - df[df["sample"] == 0]["invalid_ballots"].mean()

# Number of permutations
num_permutations = 1000
permuted_itts = []

# Perform permutation test
for _ in range(num_permutations):
    df["sample_shuffled"] = np.random.permutation(df["sample"])  # Shuffle treatment assignment
    permuted_itt = df[df["sample_shuffled"] == 1]["invalid_ballots"].mean() - df[df["sample_shuffled"] == 0]["invalid_ballots"].mean()
    permuted_itts.append(permuted_itt)

# Compute p-value
p_value = np.mean(np.abs(permuted_itts) >= np.abs(observed_itt))

# Output results
print(f"Observed ITT: {observed_itt:.4f}")
print(f"P-value from randomization inference: {p_value:.4f}")

# Interpretation
if p_value < 0.05:
    print("Reject the null hypothesis: There is evidence of an ITT effect.")
else:
    print("Fail to reject the null hypothesis: No strong evidence of an ITT effect.")


Observed ITT: 4.8241
P-value from randomization inference: 0.5030
Fail to reject the null hypothesis: No strong evidence of an ITT effect.
